### PR TITLE
Neon transactions with Solana's signature

### PR DESCRIPTION
**Pull request by anton-lisanin**

### CODE REVIEW

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11456905658/artifacts/2087011244)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   59674 |      0     |      0.162 |    0.000754868 |        0      |
| getAccountInfo                    |   54401 |      0     |      0.102 |    0.00126329  |        0      |
| getBlocks                         |   25506 |      0     |      0.133 |    0.000576061 |        0      |
| getBlock                          |   16697 |      0     |      0.027 |    0.00132323  |        0.001  |
| getMultipleAccounts               |   13675 |      0     |      0.07  |    0.00155722  |        0.001  |
| getLatestBlockhash                |   11060 |      0     |      0.07  |    0.00113391  |        0      |
| getTransaction                    |    7347 |      0     |      0.196 |    0.0089838   |        0.004  |
| getBlockTime                      |    5988 |      0     |      0.043 |    0.000375084 |        0      |
| sendTransaction                   |    5025 |      0     |      0.139 |    0.00641313  |        0.003  |
| getBalance                        |    1749 |      0     |      0.037 |    0.00142596  |        0      |
| getSignatureStatuses              |    1096 |      0     |      0.072 |    0.000974453 |        0      |
| getVersion                        |     586 |      0     |      0.088 |    0.0014198   |        0      |
| isBlockhashValid                  |     530 |      0     |      0.007 |    0.000277358 |        0      |
| simulateTransaction               |     518 |      0     |      0.003 |    0.000982625 |        0.001  |
| getSignaturesForAddress           |      56 |      0     |      0.002 |    0.00119643  |        0.001  |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.00059375  |        0.001  |
| getTokenAccountBalance            |      11 |      0     |      0.004 |    0.000727273 |        0      |
| getHealth                         |       3 |      0     |      0.001 |    0.000333333 |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getClusterNodes                   |       1 |      0.007 |      0.007 |    0.007       |        0.007  |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11515959437/artifacts/2103526403)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   64974 |          0 |      0.139 |    0.000732724 |        0      |
| getAccountInfo                    |   54790 |          0 |      0.109 |    0.00123721  |        0      |
| getBlocks                         |   26349 |          0 |      0.05  |    0.000546662 |        0      |
| getBlock                          |   16918 |          0 |      0.034 |    0.00118938  |        0.001  |
| getMultipleAccounts               |   13785 |          0 |      0.336 |    0.00172122  |        0.001  |
| getLatestBlockhash                |   11257 |          0 |      0.097 |    0.001354    |        0      |
| getTransaction                    |    7525 |          0 |      0.201 |    0.0105482   |        0.004  |
| getBlockTime                      |    5945 |          0 |      0.02  |    0.000398654 |        0      |
| sendTransaction                   |    5137 |          0 |      0.062 |    0.00760074  |        0.004  |
| getBalance                        |    1762 |          0 |      0.109 |    0.00123439  |        0      |
| getSignatureStatuses              |    1126 |          0 |      0.07  |    0.00101421  |        0      |
| getVersion                        |     589 |          0 |      0.051 |    0.0011528   |        0      |
| isBlockhashValid                  |     545 |          0 |      0.016 |    0.000377982 |        0      |
| simulateTransaction               |     518 |          0 |      0.003 |    0.000816602 |        0.001  |
| getSignaturesForAddress           |      60 |          0 |      0.182 |    0.00425     |        0.001  |
| requestAirdrop                    |      32 |          0 |      0.001 |    0.00065625  |        0.001  |
| getTokenAccountBalance            |       5 |          0 |      0.014 |    0.0028      |        0      |
| getHealth                         |       3 |          0 |      0.007 |    0.00266667  |        0.001  |
| getFirstAvailableBlock            |       2 |          0 |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |          0 |      0     |    0           |        0      |
| getClusterNodes                   |       1 |          0 |      0     |    0           |        0      |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11520300114/artifacts/2104787936)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   62913 |      0     |      0.177 |    0.00107483  |        0.001  |
| getAccountInfo                    |   54914 |      0     |      0.169 |    0.0019733   |        0.001  |
| getBlocks                         |   26801 |      0     |      0.095 |    0.000797172 |        0.001  |
| getBlock                          |   17616 |      0     |      0.17  |    0.00200074  |        0.002  |
| getMultipleAccounts               |   14096 |      0     |      0.168 |    0.00235634  |        0.001  |
| getLatestBlockhash                |   11487 |      0     |      0.142 |    0.00161147  |        0.001  |
| getTransaction                    |    7488 |      0     |      0.171 |    0.0133679   |        0.006  |
| getBlockTime                      |    6041 |      0     |      0.092 |    0.000726039 |        0      |
| sendTransaction                   |    5117 |      0     |      0.311 |    0.0091622   |        0.005  |
| getBalance                        |    1758 |      0     |      0.115 |    0.0023521   |        0.001  |
| getSignatureStatuses              |    1124 |      0     |      0.169 |    0.00191904  |        0.001  |
| getVersion                        |     587 |      0     |      0.052 |    0.00155196  |        0      |
| isBlockhashValid                  |     544 |      0     |      0.025 |    0.000476103 |        0      |
| simulateTransaction               |     518 |      0     |      0.004 |    0.00117761  |        0.001  |
| getSignaturesForAddress           |      62 |      0     |      0.023 |    0.00279032  |        0.002  |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.000875    |        0.001  |
| getTokenAccountBalance            |       4 |      0     |      0.001 |    0.0005      |        0.0005 |
| getHealth                         |       3 |      0     |      0.091 |    0.0303333   |        0      |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |        0.001  |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11522463281/artifacts/2105420172)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   60486 |      0     |      0.216 |    0.00120473  |        0.001  |
| getAccountInfo                    |   54932 |      0     |      0.143 |    0.00222078  |        0.001  |
| getBlocks                         |   24703 |      0     |      0.081 |    0.000757924 |        0.001  |
| getBlock                          |   16265 |      0     |      0.152 |    0.00197848  |        0.002  |
| getMultipleAccounts               |   13800 |      0     |      0.172 |    0.0028108   |        0.001  |
| getTransaction                    |   12598 |      0     |      0.321 |    0.0141412   |        0.005  |
| getLatestBlockhash                |   11015 |      0     |      0.213 |    0.00225828  |        0.001  |
| getBlockTime                      |    6065 |      0     |      0.196 |    0.000810552 |        0      |
| sendTransaction                   |    5146 |      0     |      0.483 |    0.011622    |        0.006  |
| getBalance                        |    1762 |      0     |      0.059 |    0.00206754  |        0.001  |
| getSignatureStatuses              |    1136 |      0     |      0.123 |    0.0019463   |        0.001  |
| getVersion                        |     598 |      0     |      0.053 |    0.00221906  |        0      |
| isBlockhashValid                  |     550 |      0     |      0.063 |    0.000785455 |        0      |
| simulateTransaction               |     518 |      0     |      0.004 |    0.00120463  |        0.001  |
| getSignaturesForAddress           |      59 |      0     |      0.017 |    0.00194915  |        0.002  |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.00084375  |        0.001  |
| getTokenAccountBalance            |       7 |      0     |      0.01  |    0.00314286  |        0.001  |
| getHealth                         |       4 |      0     |      0.001 |    0.0005      |        0.0005 |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |        0.001  |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11524575801/artifacts/2106046056)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   59676 |          0 |      0.234 |    0.00137084  |        0.001  |
| getAccountInfo                    |   55401 |          0 |      0.196 |    0.00226568  |        0.001  |
| getBlocks                         |   24028 |          0 |      0.05  |    0.000795489 |        0.001  |
| getBlock                          |   15760 |          0 |      0.134 |    0.00202538  |        0.002  |
| getMultipleAccounts               |   13884 |          0 |      0.173 |    0.00318021  |        0.001  |
| getTransaction                    |   12749 |          0 |      0.336 |    0.0147884   |        0.005  |
| getLatestBlockhash                |   10862 |          0 |      0.188 |    0.00272933  |        0.001  |
| getBlockTime                      |    6169 |          0 |      0.069 |    0.000850057 |        0      |
| sendTransaction                   |    5213 |          0 |      0.175 |    0.0131237   |        0.007  |
| getBalance                        |    1777 |          0 |      0.07  |    0.00251097  |        0.001  |
| getSignatureStatuses              |    1162 |          0 |      0.121 |    0.00237005  |        0.001  |
| getVersion                        |     615 |          0 |      0.058 |    0.00264065  |        0.001  |
| isBlockhashValid                  |     563 |          0 |      0.047 |    0.000914742 |        0      |
| simulateTransaction               |     518 |          0 |      0.004 |    0.00141699  |        0.001  |
| getSignaturesForAddress           |      58 |          0 |      0.005 |    0.00155172  |        0.002  |
| requestAirdrop                    |      32 |          0 |      0.002 |    0.000875    |        0.001  |
| getTokenAccountBalance            |       9 |          0 |      0.008 |    0.00155556  |        0.001  |
| getHealth                         |       4 |          0 |      0.001 |    0.00075     |        0.001  |
| getFirstAvailableBlock            |       2 |          0 |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |          0 |      0     |    0           |        0      |
| getClusterNodes                   |       1 |          0 |      0     |    0           |        0      |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   59453 |      0     |      0.142 |    0.000695726 |        0      |
| getAccountInfo                    |   54345 |      0     |      0.122 |    0.00137069  |        0      |
| getBlocks                         |   24374 |      0     |      0.045 |    0.000540207 |        0      |
| getBlock                          |   15749 |      0     |      0.215 |    0.0012489   |        0.001  |
| getMultipleAccounts               |   13125 |      0     |      0.12  |    0.00164968  |        0.001  |
| getTransaction                    |   12320 |      0     |      0.184 |    0.00841761  |        0.003  |
| getLatestBlockhash                |   10735 |      0     |      0.187 |    0.0013225   |        0      |
| getBlockTime                      |    5827 |      0     |      0.038 |    0.000433671 |        0      |
| sendTransaction                   |    5040 |      0     |      0.068 |    0.00630952  |        0.003  |
| getBalance                        |    1757 |      0     |      0.049 |    0.00137336  |        0      |
| getSignatureStatuses              |    1098 |      0     |      0.119 |    0.00127778  |        0      |
| getVersion                        |     578 |      0     |      0.02  |    0.000908304 |        0      |
| isBlockhashValid                  |     531 |      0     |      0.028 |    0.000423729 |        0      |
| simulateTransaction               |     518 |      0     |      0.004 |    0.000866795 |        0.001  |
| getSignaturesForAddress           |      60 |      0     |      0.003 |    0.00116667  |        0.001  |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.000625    |        0.001  |
| getTokenAccountBalance            |       6 |      0     |      0.013 |    0.00266667  |        0.0005 |
| getHealth                         |       3 |      0     |      0     |    0           |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

File: evm_loader/program/src/account/state.rs [hunk 1]
```diff
@@ -186,6 +211,11 @@ impl<'a> StateAccount<'a> {
             })
             .collect::<TreeMap<Pubkey, AccountRevision>>();
 
+        assert!(
+            !(transaction.is_scheduled_tx() ^ tree_account.is_some()),
+            "Tree account should be present iff it's a scheduled transaction."
```
Comment (s-medvedev) [inline evm_loader/program/src/account/state.rs:216]: Typo

File: evm_loader/program/src/account/transaction_tree.rs [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+            ExitStatus::Stop | ExitStatus::Suicide => (Status::Success, keccak256(&[])),
+            ExitStatus::Return(result) => (Status::Success, keccak256(result)),
+            ExitStatus::Revert(result) => (Status::Failed, keccak256(result)),
+            ExitStatus::StepLimit => {
+                panic!("Tree Account transaction can't be ended with StepLimit")
+            }
+        };
+
+        node.status = status;
+        node.result_hash = result_hash;
```
Comment (s-medvedev) [inline evm_loader/program/src/account/transaction_tree.rs:422]: Who and how use `node.result_hash`?

File: evm_loader/program/src/instruction/scheduled_transaction_start_from_account.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            let storage = StateAccount::new(
+                program_id,
+                holder,
+                &accounts_db,
+                origin,
+                trx,
+                Some(*transaction_tree.info().key),
+            )?;
+
+            do_scheduled_start(accounts_db, storage, transaction_tree, gasometer)
```
Comment (s-medvedev) [inline evm_loader/program/src/instruction/scheduled_transaction_start_from_account.rs:78]: This code the same as in scheduled_transaction_start_from_instruction.rs

**Review by s-medvedev — DISMISSED** [review]
Look ok for me

File: evm_loader/program/src/account/transaction_tree.rs (hunk 2 above)
Comment (anton-lisanin) [reply evm_loader/program/src/account/transaction_tree.rs:422]: It is not used at the moment.
Reserve for the future

File: evm_loader/program/src/instruction/scheduled_transaction_start_from_account.rs (hunk 3 above)
Comment (anton-lisanin) [reply evm_loader/program/src/instruction/scheduled_transaction_start_from_account.rs:78]: There are minor differences, but yes it is mostly the same.
Would need to be refactored later.

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/12179022615/artifacts/2278882670)

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/12184404839/artifacts/2280602393)

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/12223380254/artifacts/2290662498)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   66275 |      0     |      0.16  |    0.000502754 |         0     |
| getAccountInfo                    |   62981 |      0     |      0.238 |    0.000762405 |         0.001 |
| getBlocks                         |   28910 |      0     |      0.048 |    0.000467036 |         0     |
| getMultipleAccounts               |   15370 |      0     |      0.142 |    0.00167469  |         0.001 |
| getBlock                          |   14596 |      0     |      0.037 |    0.00124582  |         0.001 |
| getLatestBlockhash                |   10608 |      0     |      0.131 |    0.0015527   |         0.001 |
| getTransaction                    |   10353 |      0     |      0.244 |    0.0104287   |         0.003 |
| getBlockTime                      |    6285 |      0     |      0.145 |    0.000542721 |         0     |
| sendTransaction                   |    5247 |      0     |      0.119 |    0.00847036  |         0.005 |
| getBalance                        |    1800 |      0     |      0.038 |    0.000831667 |         0     |
| getSignatureStatuses              |    1146 |      0     |      0.133 |    0.00112391  |         0.001 |
| getVersion                        |     595 |      0     |      0.026 |    0.000613445 |         0     |
| isBlockhashValid                  |     555 |      0     |      0.01  |    0.000524324 |         0     |
| simulateTransaction               |     545 |      0     |      0.025 |    0.00129725  |         0.001 |
| getSignaturesForAddress           |      61 |      0.001 |      0.003 |    0.00186885  |         0.002 |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.0008125   |         0.001 |
| getHealth                         |       3 |      0.001 |      0.001 |    0.001       |         0.001 |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |         0     |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getClusterNodes                   |       1 |      0     |      0     |    0           |         0     |

</details>

**Pull request merged**